doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,592 @@
+#### 2. Window aggregations
+
+These operate on frames (a certain subset of a window partition).
+For example: Compute a moving average for each user's transaction costs on windows of 6 adjacent transactions:
+```sql
+SELECT user_id, tx_id, AVG(cost) OVER
+(PARTITION BY user_id ORDER BY timestamp ASC
+   ROWS BETWEEN 5 PRECEDING AND CURRENT ROW)
+FROM transactions;
+```
+
+There is a special case where the frame includes the entire window partition: An aggregation where the frame is both UNBOUNDED PRECEDING and UNBOUNDED FOLLOWING at the same time (or having no ORDER BY achieves a similar effect) should be transformed to a grouped aggregation + self join.
+
+In all other cases, we’ll use prefix sum, for which we need to solve two tasks:
+
+*I.* We have to find the end(s) of the interval that is the frame. I.e., we need to tell **indexes** to Prefix Sum (where the index is a value of the ORDER BY column(s), as mentioned above).
+
+*II.* We’ll need to generalize Prefix Sum to not just prefixes, but arbitrary intervals. (A prefix interval is identified by one index, a general interval is identified by two indexes.)
+
+*Solving I. for each framing mode (RANGE | GROUPS | ROWS):*
+- RANGE: this is the obvious one (but probably not the most often used): The offset is just a difference in the (sparse) “index” of the prefix sum (i.e., the ORDER BY column).
+    - Btw. we might translate some inequality self-joins to this one!
+- GROUPS: (One could say that this is probably not so often used, so no need to initially support it. However, the problem is that the solution for ROWS will probably build on this, and that is the default, so that one is often used.) We have to somehow translate the offset to a difference in the “index” of the prefix sum:
+    - Naive solution: Since this is a difference in DENSE_RANK between the current row and the ends of the interval, we could compute a DENSE_RANK, and then join with that to find the “index” (the value of the ORDER BY column). However, this would be slow, because DENSE_RANK is inherently not well incrementalizable: Small input changes can lead to lots of changes in the ranks.
+    - A better solution is to calculate a count aggregation on all ranges (with prefix sum’s `aggregate`) (let's call this `counts`), and then do a logarithmic search for the index with a nested `iterate`: We start with `(index, offset)` pairs for all the possible current elements in parallel, where the pair means that we need to move `index` down by `offset` (down when looking for the lower end of the interval, or up when looking for the upper end), i.e., we need to lower `index` while lowering `offset` to 0. Then, at every step of an `iterate`, we can use a range from `counts` on each  `(index, offset)` pair: `index` is lowered by the size of the range, and `offset` is lowered by the aggregated count of the range. We want to use the biggest such range in `counts` that doesn't make `offset` negative. We can do this by an inner `iterate`.
```

Comment:
This seems to be an important algorithmic component of the solution. Should we just include a small algorithm fragment here to go along with the description? I often found myself here trying to reconstruct this fragment in my mind when reading things like "a nested `iterate`".

Reply:
I will definitely make this part more detailed.